File: September 2022/Day 15/JAVA Code/2007. Find Original Array From Doubled Array.java
class Solution {
    public int[] findOriginalArray(int[] changed) {
        int n = changed.length;
        if(n%2==1) return new int[0];
        Arrays.sort(changed);
        HashMap<Integer,Integer> map = new HashMap<>();
        
        for(int val : changed) add(map,val);
        
        int[] ans = new int[n/2];
        int j=0;
        for(int val : changed) {
            if(val==0) {
                if(map.containsKey(0) && map.get(0)>=2) {
                    ans[j]=0;
                    j++;
                    remove(map,0);
                    remove(map,0);
                }
            } else {
                if(map.containsKey(val) && map.containsKey(val*2)) {
                    ans[j]=val;
                    j++;
                    remove(map,val);
                    remove(map,val*2);
                }
            }
        }
        if(j<n/2) return new int[0];
        return ans;
    }
    
    public static void add(HashMap<Integer,Integer> map,int key) {
        if(map.containsKey(key)) {
            map.put(key,map.get(key)+1);
        } else {
            map.put(key,1);
        }
    }
    
    public static void remove(HashMap<Integer,Integer> map,int key) {
        if(map.get(key)==1) {
            map.remove(key);
        } else {
            map.put(key,map.get(key)-1);
        }
    }
}

File: September 2022/Day 15/C++ Code/2007. Find Original Array From Doubled Array.cpp
class Solution {
public:
    vector<int> findOriginalArray(vector<int>& changed) {
        int n = changed.size();
        if(n%2==1) {
            vector<int> ans;
            return ans;
        }
        sort(changed.begin(),changed.end());
        
        vector<int> fre(100002,0);
        for(int val : changed) fre[val]++;
        
        vector<int> ans;
        for(int val : changed) {
            if(val==0) {
                if(fre[0]>=2) {
                    ans.push_back(0);
                    fre[0]-=2;
                }
                continue;
            }
            if(fre[val]>0) {
                if(val*2<100002 && fre[val*2]>0) {
                    ans.push_back(val);
                    fre[val]--;
                    fre[val*2]--;
                }
            }
        }
        if(ans.size()<n/2) {
            vector<int> a1;
            return a1;
        }
        return ans;
    }
};
